### ISSUE TITLE
deno test --reporter junit get incorrect testsuite.name  when use deno_std/testing/bdd.ts  

### ISSUE THREAD

**Issue by zuisong** [denoland/deno#20047]
```ts
//filename: test-reporter_test.ts
import { afterEach, beforeEach, it } from "https://deno.land/std@0.197.0/testing/bdd.ts";
import { assert } from "https://deno.land/std@0.197.0/assert/assert.ts";

it("test ok", ()=>{
   assert(true);
})
```

```shell
$ deno test -A   --reporter junit test-reporter_test.ts

<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="deno test" tests="1" failures="0" errors="0" time="0.094">
    <testsuite name="https://deno.land/std@0.197.0/testing/_test_suite.ts" tests="1" disabled="0" errors="0" failures="0">
        <testcase name="test ok" time="0.009">
        </testcase>
    </testsuite>
</testsuites>
```

testsuite.name is  `https://deno.land/std@0.197.0/testing/_test_suite.ts`, excpect ` test-reporter_test.ts`

**Comment by HoldUpFjord** [comment]
Hey I would love to take this issue on if it's available! I'm pretty new to deno but I'd love to contribute :)

**Comment by bartlomieju** [comment]
@HoldUpFjord sure, it's still up for grabs. Feel free to ping me if you hit a roadblock.

**Comment by HoldUpFjord** [comment]
Awesome! Thanks so much @bartlomieju

**Comment by bartlomieju** [comment]
@HoldUpFjord sure, fire away with question. 

> It seems like something in the code is generating the name as a library url vs the filename. Would you say that's correct?

So it appears that the "testsuite name" is set here:
https://github.com/denoland/deno/blob/04a259e2c8b3fdca8371f35e923bc239e8bb9ec0/cli/tools/test/reporters/junit.rs#L163

The location is actually set up here:
https://github.com/denoland/deno/blob/04a259e2c8b3fdca8371f35e923bc239e8bb9ec0/cli/js/40_testing.js#L696-L709

I recently changed that bit, because I hit similar problem when implementing `node:test` polyfill (https://github.com/denoland/deno/pull/20025), but that solution is not rock-solid - it essentially special cases "wrapper" functions coming from internal Deno code. I was thinking of a more general solution to the problem, but I haven't come up with one so far.

One quick idea is that we can look for `Symbol.for("Deno.test.location")` property in the function that's being registered as a test - that way libraries like `std/bdd.rs` could provide a proper location themselves.

**Comment by bartlomieju** [comment]
@HoldUpFjord so I would suggest to proceed as follows:
- move that bit that is looking for a location into a separate `getTestCaseLocation()` helper function
- In place where we assign location (`testDesc.location = location`) change the code to be `testDesc.location = testDesc[Symbol.for("Deno.test.location")] || getTestCaseLocation()`
- Once we land this PR, we'll adjust `std/bdd.ts` to assign proper `Symbol.for("Deno.test.location")` when the test cases are created

**Comment by HoldUpFjord** [comment]
This was super informative! Thanks for taking the time for such a long write up @bartlomieju! Just to confirm, for this PR you're wanting:
1.) The logic that looks for the location into a new helper function called `getTestCaseLocation`
2.) And altering the code that assigns the location to change to be equal to our new helper function or `testDesc[Symbol.for("Deno.test.location")] `

Then we'll open another PR for adjusting `std/bdd.ts`. Does that sound correct? As a follow up question, is there any testing I should be doing after performing the changes you've outlined? If so, how would I go about that? I checked the contributing docs and the repo's readme and didn't see anything regarding that.

**Comment by bartlomieju** [comment]
Yeah, that's correct! 

> As a follow up question, is there any testing I should be doing after performing the changes you've outlined? If so, how would I go about that? I checked the contributing docs and the repo's readme and didn't see anything regarding that.

You can alter one of the existing tests (eg. `cli/tests/testdata/test/pass.ts`) where you register a test with a function that will contain `Symbol.for("Deno.test.location")` with some fake location (`foo/bar/fizz.ts:100:200`), throw an error from that test and then in the `pass.out` assert that the location reported matches what you passed to `Symbol.for("Deno.test.location")`.

**Comment by HoldUpFjord** [comment]
Awesome! I'll get write on that then :) I'll comment if I have any more questions or any updates. Appreciate all the support.

**Comment by HoldUpFjord** [comment]
Do we want the helper function nested at the same line or declared outside of the main `test` function?

**Comment by bartlomieju** [comment]
@HoldUpFjord probably declared at the top level of the file - it would be used in several places in that file.

**Comment by HoldUpFjord** [comment]
I've been hitting a few more blockers. After clearing through a few bugs I couldn't get the cargo build step. It looks like it's an issue with cmake `--- stderr
  thread 'main' panicked at '
  failed to execute command: program not found
  is `cmake` not installed?`
  
  A Github issue stated it could be an issue with my IDE. I didn't realize VSCode and Visual Studio were different IDE's, and the docs explicitly state using Visual Studio. Trying that and hoping it resolves the build step failing.

**Comment by HoldUpFjord** [comment]
Using Visual studio seems to have fixed the above problem. Build step finishes and I can run `./target/debug/deno run cli/tests/testdata/run/002_hello.ts` succesfully, which I could not do before. 👍

**Comment by HoldUpFjord** [comment]
Ending the night stuck on this error:```
 The WPT require certain entries to be present in your C:\WINDOWS\System32\drivers\etc\hosts file. Should these be configured automatically? [y/N] y
error: Uncaught (in promise) Error: Failed to open C:\WINDOWS\System32\drivers\etc\hosts (permission error). Please run this command again with sudo, or configure the entries manually.
          throw new Error(
                ^
    at file:///C:/Users/jackm/source/repos/deno/tools/wpt.ts:102:17
    at eventLoopTick (ext:core/01_core.js:183:11)
    at async setup (file:///C:/Users/jackm/source/repos/deno/tools/wpt.ts:100:20)
    at async file:///C:/Users/jackm/source/repos/deno/tools/wpt.ts:46:5
```

I tried to manually add as suggested by the CLI and got this:```
 PS C:\Users\jackm\source\repos> cd deno
PS C:\Users\jackm\source\repos\deno>  cd test_util/wpt/
PS C:\Users\jackm\source\repos\deno\test_util\wpt>  python.exe wpt make-hosts-file | Out-File $env:SystemRoot\System32\drivers\etc\hosts -Encoding ascii -Append
Out-File : Access to the path 'C:\WINDOWS\System32\drivers\etc\hosts' is denied.
At line:1 char:35
+ ... osts-file | Out-File $env:SystemRoot\System32\drivers\etc\hosts -Enco ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OpenError: (:) [Out-File], UnauthorizedAccessException
    + FullyQualifiedErrorId : FileOpenFailure,Microsoft.PowerShell.Commands.OutFileCommand
```

**Comment by bartlomieju** [comment]
@HoldUpFjord you don't need to have WPT runner set up to work on this feature. It's enough if you're able to run `cargo test integration::test` successfully.

**Comment by HoldUpFjord** [comment]
@bartlomieju Oh that's super good to know. I can run `cargo test integration::test` successfully in the Visual Studio 2022 terminal, but opening the folder to the repo crashes Visual Studio. Is this a known error with a work around? I'll try troubleshooting on my own until I hear back.

Thanks for the help and guidance.

**Comment by bartlomieju** [comment]
Oh I have no idea why that would happen. FYI if you prefer to use `VSCode` you can do that - I believe Visual Studio was only required to install appropriate compiler toolchain.

**Comment by HoldUpFjord** [comment]
Oh interesting! I couldn't get the build phase working with `VSCode`, but it seemed to build with no friction with `Visual Studio 2022`.

**Comment by bartlomieju** [comment]
Running `cargo test integration::test` in your VSCode terminal should work if the same is working in Visual Studio.

**Comment by HoldUpFjord** [comment]
Tried it in VSCode and it didn't work, headed over to discord now :)

**Comment by HoldUpFjord** [comment]
Got some help over on Discord and got the integration tests running! This was the result: 
```
test result: FAILED. 21 passed; 55 failed; 1 ignored; 0 measured; 1265 filtered out; finished in 5.07s
error: test failed, to rerun pass `-p deno --test integration_tests```

I'll go ahead and make the changes discussed earlier for this issue and rerun them.

**Comment by HoldUpFjord** [comment]
> move that bit that is looking for a location into a separate getTestCaseLocation() helper function
In place where we assign location (testDesc.location = location) change the code to be testDesc.location = testDesc[Symbol.for("Deno.test.location")] || getTestCaseLocation()

I made the changes we referenced above for this part of the issue

**Comment by HoldUpFjord** [comment]
I feel like they're some gaps I have when working with the tests. As discussed here:
> You can alter one of the existing tests (eg. `cli/tests/testdata/test/pass.ts`) where you register a test with a function that will contain `Symbol.for("Deno.test.location")` with some fake location (`foo/bar/fizz.ts:100:200`), throw an error from that test and then in the `pass.out` assert that the location reported matches what you passed to `Symbol.for("Deno.test.location")`.

 In the `pass.ts` file, does this look like what you had in mind for the new test? : 
```
Deno.test("denoTestLocation", async () => {
  await Symbol.for("foo/bar/fizz.ts:100:200");
  throw 'error';
})
```


 Regarding the assertion, I've never worked with `.out` files. I did some leg work to learn about common use practices with assertions in Deno, including reading the docs:https://deno.land/manual@v1.36.1/basics/testing/assertions, and couldn't find an example of using it in the `.out` file. When I tried to input an assertion in said file, it wouldn't show up in the output when I ran the tests. I feel like I still have a lot of gaps in how I should proceed, as well as how everything fits together. Would you mind if I asked for some help and advice on this?

**Comment by bartlomieju** [comment]
>In the pass.ts file, does this look like what you had in mind for the new test? :
```
Deno.test("denoTestLocation", async () => {
  await Symbol.for("foo/bar/fizz.ts:100:200");
  throw 'error';
})
```

Not really, this is what I meant:
```js
const testFn = () => {
  throw new Error("boom!")
};
testFn[Symbol.for("Deno.test.location")] = "foo/bar/fizz.ts:100:200";
Deno.test("denoTestLocation", testFn);
```

>Regarding the assertion, I've never worked with .out files. I did some leg work to learn about common use practices with assertions in Deno, including reading the docs:https://deno.land/manual@v1.36.1/basics/testing/assertions, and couldn't find an example of using it in the .out file. When I tried to input an assertion in said file, it wouldn't show up in the output when I ran the tests. I feel like I still have a lot of gaps in how I should proceed, as well as how everything fits together. Would you mind if I asked for some help and advice on this?

This is really an internal API and not documented well in the contributing guide. Once you add a test like above in the `pass.out` you should add something like:

```
denoTestLocation => /foo/bar/fizz.ts:100:200
error: boom!
  throw new Error();
        ^
    at [WILDCARD]/test/fail.ts:29:9
```

I can't say specifically if that will pass (most likely not), but please open a PR and we can work more easily from there.

**Comment by HoldUpFjord** [comment]
Morning! Opened up a draft PR here:https://github.com/denoland/deno/pull/20224.

Thanks for guidance, I'll make these changes and see how the tests react.

**Comment by csvn** [comment]
Seems the PR to address this issue has stalled a bit. I made a small reproduction with some differences with Vitest's output.

1. `mkdir example && cd example`
2. Grab the code below and create `gen.ts`
3. `deno run -W gen.ts`
4. Run commands printed to generate `deno-junit.xml` & `vitest-junit.xml`

<details>

<summary>Code for generating repro</summary>

```ts
const addFile = ['add.ts', `\
export function add(a: number, b: number) {
  if (typeof a !== 'number' || typeof b !== 'number') {
    throw new Error('Both arguments must be numbers');
  }
  return a + b;
}
`];

const repro = `\
import { add } from './add.ts';


describe('add()', () => {
  it('pass', () => {});
  it.skip('skip', () => {});
  it('error', () => {
    (globalThis as any).missing();
  });
  it('failure', () => {
    expect(0).toBe(1);
  });
});\
`;

const denoTest = ['add-deno.test.ts', `\
import { describe, it } from 'jsr:@std/testing@1.0.17/bdd';
import { expect } from 'jsr:@std/expect@1.0.17';
${repro}
`];
const vitestTest = ['add-vitest.test.ts', `\
import { describe, it, expect } from 'vitest';
${repro}
`];

for (const [file, contents] of [addFile, denoTest, vitestTest]) {
  await Deno.writeTextFile('./' + file, contents);
}

console.log(`ℹ️%c  Generated reproduction files. Run commands below to generate JUnit XML files:
%c
$ deno test --junit-path=deno-junit.xml ./add-deno.test.ts
$ deno x vitest@4.0.18 --run --reporter=junit --outputFile=vitest-junit.xml ./add-vitest.test.ts
`, 'color: yellow;', 'color: #83df83;');
```

</details>

Using the JUnit XML from the Deno with [`dorny/test-reporter`](https://github.com/dorny/test-reporter) or [`EnricoMi/publish-unit-test-result-action`](https://github.com/EnricoMi/publish-unit-test-result-action) right now creates summaries that are pretty strange. 😅 

## Expected

Similar structure to Vitest:

- One `<testsuite>` with `name="./add-deno.test.ts"`
- Four `<testcase>`'s with `classname="./add-deno.test.ts"`

## Actual

- Three `<testsuite>`'s:
  - `name="./add-deno.test.ts"` 
  - `name="ext:cli/40_test.js"` (for "success" case)
  - `name="https://jsr.io/@std/testing/1.0.17/_test_suite.ts"` (for errors caught via `it()`)
- Five `<testcase>`'s with `classname` matching the suites above
  - One is a `<testcase>` that is a summary of all actual tests in `ext:cli/40_test.js` & `https://jsr.io/@std/testing/1.0.17/_test_suite.ts`, which would be better suites moving it's stats to parent `<testsuite>` (when not empty)
